### ISSUE TITLE
Error when using DiscreteUniform

### ISSUE THREAD

**Issue by springcoil** [pymc-devs/pymc#1579]
When reproducing http://isaacslavitt.com/2015/12/19/german-tank-problem-with-pymc-and-pystan/

With the newest rc4 - I run into the following ridiculous results when running `pm.summary(trace)`

```

N:

  Mean             SD               MC Error         95% HPD interval
  -------------------------------------------------------------------
  
  2687.276         2570.593         25.819           [256.000, 8376.000]

  Posterior quantiles:
  2.5            25             50             75             97.5
  |--------------|==============|==============|--------------|
  
  280.000        642.000        1643.000       4074.000       9145.000
```

These are off by a factor of 10. 

Anyone any ideas? I used the same code.

**Comment by fonnesbeck** [comment]
Yeah, that's odd. Interestingly, when you substitute `Uniform` for `DiscreteUniform` in either or both of the prior or likelihood (inappropriate, of course, but just to get an idea of whether it may be the sampler or the distribution), it does not seem to make a difference.

**Comment by twiecki** [comment]
Should probably figure that out before 3.0.

**Comment by fonnesbeck** [comment]
I guess we need to try it on different commits / release candidates and see where we broke it. Obviously it used to work.

**Comment by twiecki** [comment]
`git bisect` is perfect for this.

**Comment by twiecki** [comment]
This is the offending commit: https://github.com/pymc-devs/pymc3/commit/1b7b4bb2c5b49ea92794d86285b2c5086af7ed99

@AustinRochford Any ideas?

**Comment by twiecki** [comment]
This was part of the Mixture Distribution and relates to broadcasting, but not sure exactly what's going on there.

**Comment by fonnesbeck** [comment]
I can confirm that dropping in the old function fixes this issue.

**Comment by springcoil** [comment]
Interesting. Maybe something is going weird in the broadcasting. It might
be a good regression test to add one based on this discrete uniform.

On 7 Dec 2016 4:19 PM, "Chris Fonnesbeck" <notifications@github.com> wrote:

I can confirm that dropping in the old function fixes this issue.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/pymc-devs/pymc3/issues/1579#issuecomment-265492717>,
or mute
the thread

**Comment by fonnesbeck** [comment]
The original function does not return a True/False or 0/1, but rather the product of the `vals`, which is odd. Even putting `tt.all([ret])` causes the example to fail.

**Comment by ColCarroll** [comment]
@fonnesbeck was trying a few of these -- it looks like the function is getting a tuple of float32's (at least in this case!)

Replacing the function with, for example `return np.prod(vals)` fixes the issue (thought I didn't run the test suite).

**Comment by fonnesbeck** [comment]
@ColCarroll That makes sense, because that's exactly what the original function was doing. But, the only place `alltrue` is being used is in `bound`, where a 0/1 return value should suffice.

**Comment by AustinRochford** [comment]
@springcoil weird implicit broadcasting could definitely be the culprit.  See https://github.com/pymc-devs/pymc3/issues/1449 for the motivation for the change to `bound`.  Not sure exactly what is going on here, though.

**Comment by springcoil** [comment]
@ColCarroll Do you think you can refactor this a bit and see if the proposed change from you worked. I think we need the change from @AustinRochford because it's well motivated and just an edge case that handles this case.

**Comment by twiecki** [comment]
I think I understand.

The old `alltrue()` returned a bit-wise boolean mask which would maintain the shape of the logp.
The new `alltrue()` returns a boolean scalar which collapses the shape of the logp to be scalar as well. In this case, it's as if there's just a single data point, while in reality there are 4.

So it's clearly a bug in `alltrue()`. Next question is why `Mixture` was tripping up about it.

**Comment by twiecki** [comment]
Interestingly enough, even with the old, broadcasting `alltrue()`, the example notebook: https://github.com/pymc-devs/pymc3/blob/master/docs/source/notebooks/marginalized_gaussian_mixture_model.ipynb works just fine.

@AustinRochford do you remember why that was necessary?

**Comment by twiecki** [comment]
I actually wonder how this didn't cause problems with other likelihoods..

**Comment by AustinRochford** [comment]
@twiecki it wasn't that it didn't work for mixtures, but that `logp`s would be different based on the shape of arguments passed to `bound` (even when the likelihoods were mathematically equivalent).  

We can revert the change and look for another path forward, but changing `logp`s based on the shape passed to `bound` can cause models to converge to different points, so we will likely want to address that issue (in my opinion).

**Comment by fonnesbeck** [comment]
The old `alltrue` returned the product of `vals`, which is not necessarily boolean, unless the vals are all ones. In fact, if you change the return value to a one or zero, you get the wrong answer in the DiscreteUniform example (but a *different* wrong answer than using `tt.all`).

**Comment by fonnesbeck** [comment]
The following seems to do the trick:

```
def alltrue(vals):
    return tt.all(vals, axis=0)
```

in the sense that it combines conditions while retaining the data dimension. This  means some of the tests for `alltrue` are not what we need. For example,

```
assert alltrue([tt.ones(10)]).eval()
```

would rightly fail because the dimensions would be maintained and assert doesn't know how to deal with that.

Is there a good mixture to test this on?

**Comment by twiecki** [comment]
@AustinRochford 
> We can revert the change and look for another path forward, but changing logps based on the shape passed to bound can cause models to converge to different points, so we will likely want to address that issue (in my opinion).

I think what the previous code was trying to do is not change logps or shape if the bounds are valid. It seems you're saying though that that is what it's doing? I don't understand where `alltrue` goes wrong.

I tested @fonnesbeck's proposal and it indeed seems to do the trick, even for higher dims. Might even be more efficient. And yes, I think the test needs to change to:
`assert alltrue(vals).eval().shape == vals.shape`

While we're at it, perhaps `alltrue` could be clarified to `alltrue_mask` or `alltrue_broadcast` or something like that.

**Comment by twiecki** [comment]
OK wow, this was super confusing.

The bug is not in `alltrue` or `bound` but in `DiscreteUniform`. @AustinRochford's change converted `alltrue` to always return a scalar. That makes sense as it should not matter if some or all conditions break for the input.

However, `DiscreteUniform` implicitly depended on `alltrue` for broadcasting:
```
        return bound(-tt.log(upper - lower + 1), # only pass a scalar logp
                     lower <= value, value <= upper)
```

With Austin's change, the broadcasting wasn't done anymore by `alltrue` leading to only a single `logp` being returned by `bound`.

The proper fix should thus be:
```
        return bound(-tt.log(upper - lower + 1) * tt.ones_like(value),
                     lower <= value, value <= upper)
```

To make the passed in logp have the right shape to begin with. This fixes the problem here.

**Comment by ColCarroll** [comment]
Wow -- tricky!  

I'll look at whether the unittests could be updated to confirm that the shape of `logp` is respected, which should provide a regression test for this.

**Comment by twiecki** [comment]
Actually it's not super obvious what "the right fix" should be, it's more of a design issue. I do have the feeling that requiring logps to respect shape and alltrue not do any broadcasting is the least confusing but happy to hear other people's thoughts.

**Comment by twiecki** [comment]
CC @jsalvatier

**Comment by jsalvatier** [comment]
I think I remember purposefully making `alltrue` and `bound` precisely to avoid things like `... * tt.ones_like(value)`. 

I do tend to be too tricksy in my programming, so maybe that's a mistake. If its something we have to do all over the place, then it seems good to avoid needing to do that. 

I do think it does make some sense because 'bound' and 'alltrue' are array functions and array functions normally broadcast. Also, if its an elementwise distribution, there will *always* be at least one input into `bound` that uses `value` (so it has the right shape), so the broadcasting will work.

Maybe to address the `bound` issue, the current bound should be changed to `elem_bound` or something, since its specifically for elementwise ops. 

We definitely do need an operation that does all these things elemwise (and doesn't collapse) because the logps should be elementwise when the distribution is elementwise. That's needed for things like missing values. it also lets you inspect the logps in a more detailed way which is useful for debugging and monitoring.

**Comment by jsalvatier** [comment]
We should document the expected behavior of `bound` and `logp` (that it should be elemwise when dist is elemwise` and why) so we don't run into this in the future. 

Sorry about not documenting in the past.

**Comment by jsalvatier** [comment]
(its kinda weird to be in the position of having really detailed reasons for subtle design decisions for old systems that other people are working on).

**Comment by twiecki** [comment]
Thanks for chiming in @jsalvatier. That reasoning makes sense to me, it is a nice feature and the logp in DiscreteUniform is more elegant due to it. I also don't understand why it would lead to incorrect inference in any case.

Properly documenting it and changing the name should avoid that confusion in the future. As well as tests.

**Comment by fonnesbeck** [comment]
@twiecki that makes sense, particularly as it pertains to `DiscreteUniform` (and I assume, `Uniform`), since the uniform likelihood does not depend on the data, so the data's dimensionality does not come into play.

So, it sounds like this is only something we need to do for these distributions.

**Comment by fonnesbeck** [comment]
I can put together a PR, if someone isn't already on it.

**Comment by twiecki** [comment]
@fonnesbeck Great, not me.

**Comment by twiecki** [comment]
I did write this test:
```python
def test_bound():
    logp = tt.ones((10, 10))
    cond = tt.ones((10, 10))
    assert np.all(bound(logp, cond).eval() == logp.eval())

    logp = tt.ones((10, 10))
    cond = tt.zeros((10, 10))
    assert np.all(bound(logp, cond).eval() == (-np.inf * logp).eval())

    logp = tt.ones((10, 10))
    cond = True
    assert np.all(bound(logp, cond).eval() == logp.eval())
```

**Comment by fonnesbeck** [comment]
So, if `alltrue` ends up just being a call to `tt.all`:

    tt.all(1*vals, axis=0)

do we really need `alltrue`? The only place this is used is `bound`.

**Comment by ColCarroll** [comment]
What does multiplying by 1 do?

**Comment by fonnesbeck** [comment]
Changes booleans to integers, which sometimes avoids Theano problems (since Theano does not have a `bool` type. May not be necessary, but doesn't hurt.

**Comment by twiecki** [comment]
> do we really need alltrue? The only place this is used is bound.

I think it's a good level of abstraction, but can also change it I guess.

**Comment by fonnesbeck** [comment]
Unfortunately, the whole thing blows up when conditions are a mix of Theano and Python conditionals:

    (Elemwise{ge,no_inplace}.0, Elemwise{le,no_inplace}.0, True, True)

results in:

    theano.tensor.var.AsTensorError: ('Cannot convert (Elemwise{ge,no_inplace}.0, Elemwise{le,no_inplace}.0, True, True) to TensorType', <class 'tuple'>)

**Comment by ferrine** [comment]
We can convert all conditions to theano wrapping it with `map`.
```python
def _as_tensor(v):
    if not isinstance(v, bool):
        v = tt.as_tensor(v)
    else:
        v = tt.as_tensor(int(v))
    return v
```

**Comment by springcoil** [comment]
@ferrine that might be a good idea. Some decorator to make sure all of these conditions get mapped to Theano.

**Comment by fonnesbeck** [comment]
OK, I may have finally fixed this. Please test #1591

**Comment by twiecki** [comment]
Closed via #1596.
